File: src/Alexander/validation/swap.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <dirent.h>
#include <errno.h>
#include "qsort.c"
// External function from qsort.c
extern void swap(int *a, int *b);

char* extract_value(const char *json, const char *key) {
    char key_pattern[128];
    // Handle both quoted and unquoted keys, and potential whitespace
    snprintf(key_pattern, sizeof(key_pattern), "\"%s\"", key);
    
    const char *key_pos = strstr(json, key_pattern);
    if (!key_pos) {
        // Try without quotes around the key
        key_pos = strstr(json, key);
        if (!key_pos) return NULL;
        key_pos += strlen(key);
    } else {
        key_pos += strlen(key_pattern);
    }
    
    // Skip whitespace and colon
    while (*key_pos && (isspace(*key_pos) || *key_pos == ':')) {
        key_pos++;
    }
    
    // Expect opening quote
    if (*key_pos != '"') return NULL;
    key_pos++; // Skip opening quote
    
    const char *val_start = key_pos;
    const char *val_end = strchr(val_start, '"');
    if (!val_end) return NULL;
    
    size_t len = val_end - val_start;
    char *result = malloc(len + 1);
    if (!result) return NULL;
    
    memcpy(result, val_start, len);
    result[len] = '\0';
    return result;
}

unsigned char* parse_bytes(const char *str, size_t *byte_count) {
    // Handle format: "b'\\x00\\x01...'"
    const char *ptr = str;
    
    if (strncmp(ptr, "b'", 2) == 0 && ptr[strlen(ptr)-1] == '\'') {
        ptr += 2;  // Skip "b'"
    }
    
    // Count how many \x sequences we have
    size_t count = 0;
    const char *temp = ptr;
    while (*temp && *temp != '\'') {
        if (*temp == '\\' && *(temp + 1) == 'x') {
            count++;
            temp += 4;  // Skip \xHH
        } else {
            temp++;
        }
    }
    
    *byte_count = count;
    unsigned char *bytes = malloc(*byte_count);
    if (!bytes) return NULL;
    
    size_t i = 0;
    while (*ptr && *ptr != '\'' && i < count) {
        if (*ptr == '\\' && *(ptr + 1) == 'x') {
            // Parse hex digits
            char hex_str[3] = {ptr[2], ptr[3], '\0'};
            char *endptr;
            bytes[i] = (unsigned char)strtoul(hex_str, &endptr, 16);
            
            if (*endptr != '\0') {
                fprintf(stderr, "Invalid hex digits: %s\n", hex_str);
                free(bytes);
                return NULL;
            }
            i++;
            ptr += 4;  // Skip \xHH
        } else {
            ptr++;
        }
    }
    
    return bytes;
}

int main() {
    const char *dir_path = "data_out_swap";
    DIR *dir = opendir(dir_path);
    if (!dir) {
        fprintf(stderr, "Error opening directory '%s': %s\n", dir_path, strerror(errno));
        return EXIT_FAILURE;
    }

    struct dirent *entry;
    while ((entry = readdir(dir)) != NULL) {
        if (strcmp(entry->d_name, ".") == 0 || strcmp(entry->d_name, "..") == 0)
            continue;
            
        char filepath[1024];
        snprintf(filepath, sizeof(filepath), "%s/%s", dir_path, entry->d_name);
        
        FILE *f = fopen(filepath, "r");
        if (!f) {
            fprintf(stderr, "Error opening file '%s': %s\n", filepath, strerror(errno));
            continue;
        }

        fseek(f, 0, SEEK_END);
        long size = ftell(f);
        fseek(f, 0, SEEK_SET);
        
        char *content = malloc(size + 1);
        if (!content) {
            fclose(f);
            continue;
        }
        
        fread(content, 1, size, f);
        fclose(f);
        content[size] = '\0';
        
        char *a_str = extract_value(content, "a");
        char *b_str = extract_value(content, "b");
        
        if (!a_str || !b_str) {
            fprintf(stderr, "JSON extraction failed in %s\n", filepath);
            free(a_str);
            free(b_str);
            free(content);
            continue;
        }
        
        size_t a_bytes, b_bytes;
        unsigned char *a_bytes_arr = parse_bytes(a_str, &a_bytes);
        unsigned char *b_bytes_arr = parse_bytes(b_str, &b_bytes);
        
        free(a_str);
        free(b_str);
        free(content);
        
        if (!a_bytes_arr || !b_bytes_arr) {
            fprintf(stderr, "Byte parsing failed in %s\n", filepath);
            free(a_bytes_arr);
            free(b_bytes_arr);
            continue;
        }

        // Convert byte arrays to integer arrays (little-endian)
        // For swap, we expect each to be an integer array
        if (a_bytes % sizeof(int) != 0 || b_bytes % sizeof(int) != 0) {
            fprintf(stderr, "Invalid byte size: a=%zu, b=%zu (not divisible by %zu)\n", 
                    a_bytes, b_bytes, sizeof(int));
            free(a_bytes_arr);
            free(b_bytes_arr);
            continue;
        }
        
        int a_len = a_bytes / sizeof(int);
        int b_len = b_bytes / sizeof(int);
        int *a_arr = (int *)a_bytes_arr;
        int *b_arr = (int *)b_bytes_arr;
        
        printf("Input a: [");
        for (int i = 0; i < a_len; i++) {
            printf("%d%s", a_arr[i], (i < a_len - 1) ? ", " : "");
        }
        printf("]\n");
        
        printf("Input b: [");
        for (int i = 0; i < b_len; i++) {
            printf("%d%s", b_arr[i], (i < b_len - 1) ? ", " : "");
        }
        printf("]\n");
        
        // Call swap on each corresponding pair of elements
        int min_len = (a_len < b_len) ? a_len : b_len;
        for (int i = 0; i < min_len; i++) {
            swap(&a_arr[i], &b_arr[i]);
        }
        
        printf("Output a: [");
        for (int i = 0; i < a_len; i++) {
            printf("%d%s", a_arr[i], (i < a_len - 1) ? ", " : "");
        }
        printf("]\n");
        
        printf("Output b: [");
        for (int i = 0; i < b_len; i++) {
            printf("%d%s", b_arr[i], (i < b_len - 1) ? ", " : "");
        }
        printf("]\n");
        
        free(a_bytes_arr);
        free(b_bytes_arr);
    }
    
    closedir(dir);
    return EXIT_SUCCESS;
}

File: src/Alexander/validation/swap.rs
use std::fs;

mod qsort;

// Swap function from qsort.rs module
use qsort::swap;

fn extract_value(json: &str, key: &str) -> Option<String> {
    let key_pattern = format!("\"{}\"", key);
    
    let key_pos = if let Some(pos) = json.find(&key_pattern) {
        pos + key_pattern.len()
    } else if let Some(pos) = json.find(key) {
        pos + key.len()
    } else {
        return None;
    };
    
    let remaining = &json[key_pos..];
    
    // Skip whitespace and colon
    let mut chars = remaining.chars();
    let mut start_pos = 0;
    for ch in chars.by_ref() {
        if ch == ':' || ch.is_whitespace() {
            start_pos += ch.len_utf8();
        } else {
            break;
        }
    }
    
    let value_part = &remaining[start_pos..];
    
    // Expect opening quote
    if !value_part.starts_with('"') {
        return None;
    }
    
    let value_part = &value_part[1..]; // Skip opening quote
    
    if let Some(end_pos) = value_part.find('"') {
        Some(value_part[..end_pos].to_string())
    } else {
        None
    }
}

fn parse_bytes(s: &str) -> Option<Vec<u8>> {
    let mut ptr = s;
    
    // Handle format: "b'\\x00\\x01...'"
    if ptr.starts_with("b'") && ptr.ends_with('\'') {
        ptr = &ptr[2..ptr.len()-1]; // Remove "b'" and trailing "'"
    }
    
    let mut bytes = Vec::new();
    let mut chars = ptr.chars().peekable();
    
    while let Some(ch) = chars.next() {
        if ch == '\\' && chars.peek() == Some(&'x') {
            chars.next(); // consume 'x'
            
            // Get next two hex digits
            let hex1 = chars.next()?;
            let hex2 = chars.next()?;
            
            let hex_str = format!("{}{}", hex1, hex2);
            let byte_val = u8::from_str_radix(&hex_str, 16).ok()?;
            bytes.push(byte_val);
        }
    }
    
    Some(bytes)
}

fn bytes_to_i32_array_le(bytes: &[u8]) -> Vec<i32> {
    bytes.chunks_exact(4)
        .map(|chunk| {
            i32::from_le_bytes([chunk[0], chunk[1], chunk[2], chunk[3]])
        })
        .collect()
}

fn main() -> std::io::Result<()> {
    let dir_path = "data_out_swap";
    let entries = fs::read_dir(dir_path)?;
    
    for entry in entries {
        let entry = entry?;
        let path = entry.path();
        
        if path.is_file() {
            let content = fs::read_to_string(&path)?;
            
            let a_str = extract_value(&content, "a");
            let b_str = extract_value(&content, "b");
            
            if let (Some(a_str), Some(b_str)) = (a_str, b_str) {
                let a_bytes = parse_bytes(&a_str);
                let b_bytes = parse_bytes(&b_str);
                
                if let (Some(a_bytes), Some(b_bytes)) = (a_bytes, b_bytes) {
                    let mut a_arr = bytes_to_i32_array_le(&a_bytes);
                    let mut b_arr = bytes_to_i32_array_le(&b_bytes);
                    
                    print!("Input a: [");
                    for (i, &val) in a_arr.iter().enumerate() {
                        if i > 0 { print!(", "); }
                        print!("{}", val);
                    }
                    println!("]");
                    
                    print!("Input b: [");
                    for (i, &val) in b_arr.iter().enumerate() {
                        if i > 0 { print!(", "); }
                        print!("{}", val);
                    }
                    println!("]");
                    
                    // Call swap on each corresponding pair of elements
                    let min_len = a_arr.len().min(b_arr.len());
                    for i in 0..min_len {
                        swap(Some(&mut a_arr[i]), Some(&mut b_arr[i]));
                    }
                    
                    print!("Output a: [");
                    for (i, &val) in a_arr.iter().enumerate() {
                        if i > 0 { print!(", "); }
                        print!("{}", val);
                    }
                    println!("]");
                    
                    print!("Output b: [");
                    for (i, &val) in b_arr.iter().enumerate() {
                        if i > 0 { print!(", "); }
                        print!("{}", val);
                    }
                    println!("]");
                }
            }
        }
    }
    
    Ok(())
}
